unit_tests/test_redis_aggregator.py

```diff
@@ -0,0 +1,238 @@
+    TEST_DEVICE_UUID_1: [TEST_BATCH_COMMAND],
+    TEST_DEVICE_UUID_2: [TEST_BATCH_COMMAND]
+}
+
+TEST_TRANSACTION_ID_RESPONSE = "TEST_TRANSACTION_ID_RESPONSE"
+TEST_TRANSACTION_ID_BUFFER = {
+    TEST_TRANSACTION_ID: {"transaction_id": TEST_TRANSACTION_ID,
+                          "response": TEST_TRANSACTION_ID_RESPONSE}
+}
+
+TEST_RESPONSE = "TEST_RESPONSE"
+TEST_TRANSACTION_ID_RESPONSE_BUFFER = {TEST_TRANSACTION_ID: TEST_RESPONSE}
+
+
+@pytest.fixture(autouse=True)
+def fixture_mock_connections(mocker):
+    mocker.patch("gsy_e_sdk.redis_aggregator.StrictRedis")
+    mocker.patch("gsy_e_sdk.redis_aggregator.ThreadPoolExecutor")
+
+
+@pytest.fixture(name="mock_grid_fee_calculation")
+def fixture_mock_grid_fee_calculation(mocker):
+    """Mock GridFeeCalculation class."""
+    mocker.patch("gsy_e_sdk.redis_aggregator.GridFeeCalculation")
+
+
+@pytest.fixture(name="mock_transaction_id_and_timeout_blocking")
+def fixture_mock_transaction_id_and_timeout_blocking(mocker):
+    mocker.patch("gsy_e_sdk.redis_aggregator.uuid.uuid4",
+                 return_value=TEST_TRANSACTION_ID)
+    mocker.patch("gsy_e_sdk.redis_aggregator.wait_until_timeout_blocking")
+
+
+@pytest.fixture(name="mock_client_command_buffer_attributes")
+def fixture_mock_client_command_buffer(mocker):
+    mocker.patch("gsy_e_sdk.redis_aggregator.ClientCommandBuffer.buffer_length",
+                 new_callable=PropertyMock, return_value=3)
+    mocker.patch("gsy_e_sdk.redis_aggregator.ClientCommandBuffer.execute_batch",
+                 return_value=TEST_BATCH_COMMAND_DICT)
+
+
+@pytest.fixture(name="aggregator")
+def fixture_aggregator(mocker):
+    mocker.patch("gsy_e_sdk.redis_aggregator.RedisAggregator._create_aggregator")
+    return RedisAggregator(aggregator_name=TEST_AGGREGATOR_NAME)
+
+
+class TestRedisAggregatorConstructor:
+    """Includes tests for the constructor of the RedisAggregator class."""
+
+    @staticmethod
+    @pytest.mark.usefixtures("mock_grid_fee_calculation")
+    def test_constructor_grid_fee_calculation_instantiated(aggregator):
+        assert aggregator.grid_fee_calculation is not None
+
+    @staticmethod
+    def test_constructor_redis_db_instantiated(aggregator):
+        assert aggregator.redis_db is not None
+
+    @staticmethod
+    def test_constructor_pubsub_instantiated(aggregator):
+        assert aggregator.pubsub is not None
+
+    @staticmethod
+    def test_constructor_client_command_buffer_instantiated(aggregator):
+        assert aggregator._client_command_buffer is not None
+
+    @staticmethod
+    def test_constructor_threadpoolexecutor_instantiated(aggregator):
+        assert aggregator.executor is not None
+
+    @staticmethod
+    def test_constructor_connect_and_subscribe_called():
+        with patch("gsy_e_sdk.redis_aggregator."
+                   "RedisAggregator._connect_and_subscribe") as mocked_method:
+            RedisAggregator(aggregator_name=TEST_AGGREGATOR_NAME)
+            mocked_method.assert_called()
```

Comment:
One more thing, even though the `_subscribe_to_response_channels1 is a private method and we tend to avoid testing these methods, in this particular case I find it quite important to test it to make sure that the change in channels name is watched  (In theory I would assume that the class should keep track of its channels in a special variable that is tested)

Reply:
I agree with Hassan that here we're not really testing the effects of this method. `_connect_and_subscribe` is a private method, so I wouldn't mock it if that can be avoided. We can ignore also the internal calls to `_subscribe_to_aggregator_response_and_start_redis_thread()` and so on, because they are also private methods which would bind us to the implementation.

What we want to actually test are the side-effects of this method, pretending that these intermediate private method calls don't exist (as if everything was implemented inside of the method). What are these side-effects? For example:
1. `self.pubsub.psubscribe` is called with some specific arguments
2. `self.pubsub.run_in_thread` is called with some specific arguments
3. if `self.aggregator_uuid` is specified, then `self.aggregator_uuid` will be equal to some specific UUID
And so on.

Sidenote: the more nested methods we have, the more difficult it is to follow the path to their side-effects. This is usually a warning for us: probably we could refactor the class and split it into several classes, which would then have their own public (and testable) methods.

Reply:
I see your point and finally understand it. 
I have been working towards a cleaner implementation of the tests. I stoped mocking these private methods and making assertions about the side effects due to them. I will push those changes later when I end up addressing all of the present comments.